### PR TITLE
Add a way to configure `TypeFactory` Jackson uses

### PR DESCRIPTION
**Pull request by JooHyukKim**
(part of #2502)

~~Blocked by #4111, which has implementations in cleaner form that can be retrofitted to this PR.~~

### CODE REVIEW

File: src/test/java/com/fasterxml/jackson/databind/cfg/CacheProviderTest.java [hunk 1]
```diff
@@ -205,10 +249,12 @@ public void testBuilderValueValidation() throws Exception
         DefaultCacheProvider.builder()
                 .maxDeserializerCacheSize(0)
                 .maxSerializerCacheSize(0)
+                .maxTypeFactoryCacheSize(0)
```
Comment (cowtowncoder) [inline src/test/java/com/fasterxml/jackson/databind/cfg/CacheProviderTest.java:252]: One thing I wonder: I think that most cache implementations impose a minimum size, so `0` (for example) won't do what user might want -- that is, disabling of caching.

Then again, it is now possibly to add custom cache instances too which can be "no-op" (do nothing) so maybe this doesn't matter.

I also haven't tested out if backing `LRUMap` honors 0 size or not: JDK Maps do have minimum size but since 2.14 `LRUMap` uses integrated version of a more optimized caching library (name of which escapes me now).

Comment (iProdigy) [reply src/test/java/com/fasterxml/jackson/databind/cfg/CacheProviderTest.java:252]: `LRUMap` *does* honor max size of 0 (a write occurs, but is immediately evicted). `LRUMap` is based off [ConcurrentLinkedHashMap](https://github.com/ben-manes/concurrentlinkedhashmap), which is a predecessor to [Caffeine](https://github.com/ben-manes/caffeine)

Caffeine, on the other hand, batches evictions, so the zero maximum size constraint can be temporarily violated.

In addition, Cache2k, Android LruCache, and Ehcache all throw IllegalArgumentException when trying to create a cache with a maximum size of zero.

Comment (cowtowncoder) [reply src/test/java/com/fasterxml/jackson/databind/cfg/CacheProviderTest.java:252]: Thank you @iProdigy . Sounds like we are good then. So 0 size for 2.14+ Jackson implementation will not retain anything in cache, and while there is write overhead that should be negligible.

It'd be nice to test this invariant, but no further work should be necessary it sounds.

File: src/main/java/com/fasterxml/jackson/databind/cfg/DefaultCacheProvider.java [hunk 2]
```diff
@@ -163,13 +189,32 @@ public Builder maxSerializerCacheSize(int maxSerializerCacheSize) {
             return this;
         }
 
+        /**
+         * Define the maximum size of the {@link LookupCache} instance constructed by {@link #forTypeFactory()}
+         * and {@link #_buildCache(int)}
+         * <p>
+         * Note that value zero is not considered a magic number, unlike some other libraries where it means "disabling cache".
+         * Instead setting value to zero only means setting cache size to zero.
```
[the remark this answers is not in the document]
Comment (iProdigy) [reply src/main/java/com/fasterxml/jackson/databind/cfg/DefaultCacheProvider.java:197]: I would rephrase... For DefaultCacheProvider, zero max size effectively does disable caching since entries are not retained in the underlying LRUMap. Yes, it's technically true that a write & removal occurs under the hood, but users don't care about this detail (other than for optimizing performance, as a special no-op implementation could've been used instead)

Perhaps: `Specifying a maximum size of zero prevents reuse of values that would have been cached`

File: src/main/java/com/fasterxml/jackson/databind/type/TypeFactory.java [hunk 3]
```diff
@@ -260,11 +270,23 @@ public TypeFactory withCache(LRUMap<Object,JavaType> cache)  {
...
         return new TypeFactory(cache, _parser, _modifiers, _classLoader);
     }
 
+    /**
+     * Mutant factory method that will construct a new {@link TypeFactory}
+     * with cache instances provided by {@link CacheProvider}.
+     *
+     * @since 2.16
+     */
+    public TypeFactory withCaches(CacheProvider cacheProvider) {
```
Comment (cowtowncoder) [inline src/main/java/com/fasterxml/jackson/databind/type/TypeFactory.java:285]: I think I'll change this so we won't need a new method -- caller can just construct `LookupCache` needed and pass that.
